### PR TITLE
Add offline entitlements integration tests

### PR DESCRIPTION
**Pull request by tonidero**
### Description
Depends on #987 

This PR adds some integration tests checking some basic behaviors of offline entitlements. Mostly making sure the mode is turn on and off as expected.

### CODE REVIEW

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import io.mockk.slot
+import org.assertj.core.api.Assertions
+import org.junit.Rule
+import java.net.URL
+import java.util.Date
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import kotlin.time.Duration.Companion.seconds
+
+open class BasePurchasesIntegrationTest {
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt:19]: I moved some logic from `PurchasesIntegrationTest` to this base class.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    protected var latestPurchasesUpdatedListener: BillingAbstract.PurchasesUpdatedListener? = null
+    protected var latestStateListener: BillingAbstract.StateListener? = null
+
+    // region helpers
+
+    protected fun setupTest(
+        initialSharedPreferences: Map<String, String> = emptyMap(),
+        initialActivePurchases: Map<String, StoreTransaction> = emptyMap(),
+        postSetupTestCallback: (MainActivity) -> Unit = {}
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt:39]: This was added because `onActivityReady` wasn't guaranteed to be executed in order which resulted in some flaky tests. This way, we execute the test setup code and the test itself within a single `onActivityReady`.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    protected fun tearDownTest() {
+        Purchases.sharedInstance.close()
+    }
+
+    protected fun onActivityReady(block: (MainActivity) -> Unit) {
+        activityScenarioRule.scenario.onActivity(block)
+    }
+
+    protected fun ensureBlockFinishes(block: (CountDownLatch) -> Unit) {
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt:77]: Copied this from the `BackendIntegrationTest` which are actually unit tests unlike this.

File: purchases/src/integrationTest/AndroidManifest.xml [hunk 4]
```diff
@@ -10,6 +10,7 @@
             android:supportsRtl="true"
             android:debuggable="true"
             android:theme="@style/Theme.MaterialComponents.DayNight.DarkActionBar"
+            android:usesCleartextTraffic="true"
```
Comment (tonidero) [inline purchases/src/integrationTest/AndroidManifest.xml:13]: Had to add this to be able to use `MockWebServer` pointing to localhost.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchasesIntegrationTest.kt [hunk 5]
```diff
@@ -43,52 +33,19 @@ class PurchasesIntegrationTest {
...
-            }
-
-            Purchases.configure(
-                PurchasesConfiguration.Builder(it, Constants.apiKey)
-                    .appUserID(testUserId)
-                    .build(),
-                mockBillingAbstract
-            )
-        }
+        setupTest()
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchasesIntegrationTest.kt:43]: Ideally we should also migrate these tests to use a single `onActivityReady`. But will do that in a different PR.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+            "original_purchase_date": null,
+            "other_purchases": {},
+            "subscriptions": {}
+          },
+          "schema_version": 3,
+          "customer_info_request_date": 1684245988766
+        }
+    """.trimIndent()
+    private val cacheKeyPrefix = "com.revenuecat.purchases.${Constants.apiKey}"
+    private val productEntitlementMappingCacheKey = "$cacheKeyPrefix.productEntitlementMapping"
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:63]: It's not great to duplicate these cache keys... But I didn't want to expose them in the `common` module just for this... I think it's ok? Lmk what you think

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        val storeProduct = StoreProductFactory.createGoogleStoreProduct(
+            productId = testProductId,
+            basePlanId = testBasePlanId,
+            subscriptionOptionsList = listOf(
+                StoreProductFactory.createGoogleSubscriptionOption(
+                    productId = testProductId,
+                    basePlanId = testBasePlanId,
+                )
+            )
+        )
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:140]: Creating these fake products is a bit hacky... We could create them from mocked google product details but that also doesn't feel great to me though, I think this is relatively simpler.

**Comment by tonidero** [comment]
Still want to write some more tests including:
- Does not enter offline entitlements if it's a non-500 error
- Recovers from offline entitlements if either a getCustomerInfo or a postReceipt request succeed
- ...

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+    // region helpers
+
+    private fun setupMockServer() {
+        server = MockWebServer()
+        server.start()
+        Purchases.proxyURL = server.url("").toUrl()
+        server.dispatcher = object : Dispatcher() {
+            override fun dispatch(request: RecordedRequest): MockResponse {
+                return MockResponse().setResponseCode(RCHTTPStatusCodes.ERROR)
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:186]: Right now this is always returning 500. The plan would be to change this to also return successes in some endpoints for some tests to test recovery.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    protected lateinit var mockBillingAbstract: BillingAbstract
+
+    protected var latestPurchasesUpdatedListener: BillingAbstract.PurchasesUpdatedListener? = null
+    protected var latestStateListener: BillingAbstract.StateListener? = null
+
+    // region helpers
+
+    protected fun setupTest(
+        initialSharedPreferences: Map<String, String> = emptyMap(),
+        initialActivePurchases: Map<String, StoreTransaction> = emptyMap(),
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt:38]: Added these to be able to mock the initial state of shared preferences and active purchases within the integration test.

**Review by tonidero — COMMENTED** [review]
This is still WIP but I would like some feedback on the approach @RevenueCat/coresdk . 

I think this is much simpler than having to run a separate process for the proxy and enable/disable different endpoints in the middle of the test. On the other hand, I agree that hardcoding the success responses makes it much more brittle. I believe that it's somewhat acceptable to hardcode the responses, since we are hitting the backend in other integration tests. We can still use the proxy for manual testing though. Lmk what you think!

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineEntitlementsManager.kt [hunk 10]
```diff
@@ -78,18 +78,22 @@ class OfflineEntitlementsManager(
         )
     }
 
-    fun updateProductEntitlementMappingCacheIfStale() {
+    fun updateProductEntitlementMappingCacheIfStale(completion: (() -> Unit)? = null) {
```
Comment (tonidero) [inline common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineEntitlementsManager.kt:81]: Added a completion here to be able to wait upon it on tests. It won't be used in production. Still it's tested.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+        private fun canRunIntegrationTests() = Constants.apiKey != "REVENUECAT_API_KEY" &&
+            Constants.googlePurchaseToken != "GOOGLE_PURCHASE_TOKEN" &&
+            Constants.productIdToPurchase != "PRODUCT_ID_TO_PURCHASE"
+    }
+
+    @get:Rule
+    var activityScenarioRule = activityScenarioRule<MainActivity>()
+
+    protected val testTimeout = 7.seconds
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt:39]: Had to increase the timeout... Honestly this is already large, but in Android we do all requests sequentially and that's taking a long time for the new tests. Also, randomly the tests will take twice as long... We will probably want to improve the performance of the integration tests soon, but for now I wanted to get them running first. Also, the fact that we are using callbacks and not coroutines makes it necessary for us to put a timeout for the whole test :(

**Review by tonidero — COMMENTED** [review]
These tests don't look nearly as nice in Android as they do in iOS. One of the main differences is coroutine support. That would make the tests much more legible. I tried to avoid as many callbacks as possible but still not nearly as nice. I guess another reason to consider supporting coroutines soon

**Comment by tonidero** [comment]
This is ready for review @RevenueCat/coresdk

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+import org.assertj.core.api.Assertions.assertThat
+import org.assertj.core.api.Assertions.fail
+import org.junit.After
+import org.junit.Test
+import org.junit.runner.RunWith
+import kotlin.time.Duration
+
+@Suppress("TooManyFunctions")
+@RunWith(AndroidJUnit4::class)
+class OfflineEntitlementsIntegrationTest : BasePurchasesIntegrationTest() {
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:21]: Note that I didn't add automated tests that tested offline entitlements with multiple purchases. Reasons being:
- I would need to assume many things about how Google works in order to mock it realistically. I really wish Google released something to test purchases more easily :(.
- We would need another valid token for a different product if we wanted to test it against the backend

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+        }
+    }
+
+    private fun assertCustomerInfoDoesNotHavePurchaseData(customerInfo: CustomerInfo) {
+        assertThat(customerInfo.entitlements.active).isEmpty()
+        assertThat(customerInfo.activeSubscriptions).isEmpty()
+    }
+
+    private fun assertCustomerInfoHasExpectedPurchaseData(customerInfo: CustomerInfo) {
+        assertThat(customerInfo.entitlements.active.keys).containsExactlyInAnyOrderElementsOf(
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:373]: I wanted to note that this only works for these tests that are using offline entitlements. The token we currently are using for integration tests is for a expired purchase, that means that the backend won't return it as an active purchase. For these tests I'm mocking the purchase as returned from the Google method "queryPurchases" which returns active purchases, that's why it ends up as an active purchase there.

That's also the reason why I'm not verifying that the purchases are active after recovering from offline entitlements mode and only verifying that the purchase is acknowledged and sent to the backend.

This is not really great. I think we can improve this a lot by having an active purchase forever in a company credit card. Will study this option and try to improve this a bit.

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineEntitlementsManager.kt (hunk 10 above)
Comment (NachoSoto) [reply common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineEntitlementsManager.kt:81]: As an improvement you could forward the error too so integration tests can ensure it actually worked?

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt (hunk 11 above)
Comment (NachoSoto) [reply purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt:39]: iOS integration tests are also not the fastest because we need to wait on purchase expirations. I think it's okay that these tests aren't the fastest, as long as unit tests are.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+    protected var latestStateListener: BillingAbstract.StateListener? = null
+
+    protected val entitlementsToVerify = Constants.activeEntitlementIdsToVerify
+        .split(",")
+        .map { it.trim() }
+        .filter { it.isNotEmpty() }
+
+    // region helpers
+
+    protected fun setupTest(
```
Comment (NachoSoto) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt:56]: Nit: `set up` should be a verb here which is 2 words
```suggestion
    protected fun setUpTest(
```

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    private fun clearAllSharedPreferences(context: Context) {
+        PreferenceManager.getDefaultSharedPreferences(context).edit().clear().commit()
+        context.getSharedPreferences(
+            "${context.packageName}_preferences_etags",
+            Context.MODE_PRIVATE
+        ).edit().clear().commit()
+        context.getSharedPreferences(
+            "com_revenuecat_purchases_${context.packageName}_preferences_diagnostics",
```
Comment (NachoSoto) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt:131]: It would be nice to not hardcode here these 2 magic strings.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 16]
```diff
@@ -0,0 +1,394 @@
...
+        skus = listOf(Constants.productIdToPurchase),
+        purchaseToken = Constants.googlePurchaseToken
+    )
+    private val initialActivePurchases = mapOf(
+        initialActiveTransaction.purchaseToken.sha1() to initialActiveTransaction
+    )
+
+    @After
+    fun tearDown() {
+        tearDownTest()
```
Comment (NachoSoto) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:33]: Would it be possible to put `@After` on the parent `tearDownTest` so subclasses don't have to remember this?

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 17]
```diff
@@ -0,0 +1,394 @@
...
+                    }
+                )
+            }
+        }
+    }
+
+    @Test
+    fun entersOfflineEntitlementsModeIfNoCachedCustomerInfoAndCustomerInfoRequestReturns500() {
+        ensureBlockFinishes { latch ->
+            setupTestWaitingForInitialRequests(
```
Comment (NachoSoto) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:60]: To simplify these, I'd recommend combining tests that share the same set up (`setupTest` / `setupTestWaitingForInitialRequests`) into separate classes, so this can be done in the `@Before`. 
I try to follow that pattern to make the tests themselves easier to follow so they only focus on what they have to do once the SDK is already initialized (in different ways).
What do you think?

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+                Purchases.sharedInstance.updatedCustomerInfoListener = UpdatedCustomerInfoListener {
+                    receivedCustomerInfosInListener.add(it)
+                }
+                mockPurchaseResult()
+                mockBillingAbstract.mockQueryProductDetails(queryProductDetailsSubsReturn = listOf(storeProduct))
+                Purchases.sharedInstance.purchaseWith(
+                    PurchaseParams.Builder(activity, storeProduct).build(),
+                    onError = { error, _ ->
+                        latch.countDown()
+                        fail("Expected success but got error: $error")
```
Comment (NachoSoto) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:134]: Is it a problem that this is in this order? The code waiting for the latch would be notified before you fail the test.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+                initialActivePurchases = emptyMap()
+            ) { activity ->
+                Purchases.sharedInstance.forceServerErrors = true
+
+                mockPurchaseResult(activePurchases = mapOf(inAppTransaction.purchaseToken.sha1() to inAppTransaction))
+                mockBillingAbstract.mockQueryProductDetails(queryProductDetailsInAppReturn = listOf(inAppProduct))
+                Purchases.sharedInstance.purchaseWith(
+                    PurchaseParams.Builder(activity, inAppProduct).build(),
+                    onError = { _, _ ->
+                        latch.countDown()
```
Comment (NachoSoto) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:168]: Maybe verify the error is a server error and not something else?

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 20]
```diff
@@ -0,0 +1,394 @@
...
+            ) { activity ->
+                Purchases.sharedInstance.forceServerErrors = true
+
+                mockPurchaseResult()
+                mockBillingAbstract.mockQueryProductDetails(queryProductDetailsSubsReturn = listOf(storeProduct))
+                Purchases.sharedInstance.purchaseWith(
+                    PurchaseParams.Builder(activity, storeProduct).build(),
+                    onError = { error, _ ->
+                        latch.countDown()
+                        fail("Expected success but got error: $error")
```
Comment (NachoSoto) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:195]: Ditto order and in the other tests.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 21]
```diff
@@ -0,0 +1,394 @@
...
+                        )
+                    }
+                )
+            }
+        }
+    }
+
+    // region helpers
+
+    private fun setupTestWaitingForInitialRequests(
```
Comment (NachoSoto) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:334]: Nit:
```suggestion
    private fun setUpTestWaitingForInitialRequests(
```

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt [hunk 22]
```diff
@@ -0,0 +1,394 @@
...
+        initialActivePurchases: Map<String, StoreTransaction> = emptyMap(),
+        postSetupTestCallback: (MainActivity) -> Unit = {}
```
Comment (NachoSoto) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:336]: This is a noun, so `postSetup` makes sense 😅

**Review by NachoSoto — APPROVED** [review]
This is a great start 👍🏻

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt (hunk 17 above)
Comment (tonidero) [reply purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:60]: Well, with async/coroutines that would be much simpler, but we need to wait to run the test until the callback is received here, so we can't move it to the `@Before` easily... I guess we could create a latch to wait for that in the `@Before`. We also need the `activity` in some of these, though I guess we could cache it ourselves... Will try that to try to simplify these tests.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt (hunk 18 above)
Comment (tonidero) [reply purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt:134]: Right, I think we should completely remove the countdown in these error cases, I added it on an early iteration but it's not really needed, and can cause errors as you said. Good catch!

**Comment by tonidero** [comment]
Thanks for the suggestions @NachoSoto! Refactored the integration tests pretty heavily into different classes. So will need another review 🙏

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/offlineentitlements/OfflineEntitlementsFreshInstallIntegrationTest.kt [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+    override val initialForceServerErrors = true
+
+    @Before
+    fun setUp() {
+        ensureBlockFinishes { latch ->
+            setUpTest {
+                latch.countDown()
+            }
+        }
+    }
```
Comment (NachoSoto) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/offlineentitlements/OfflineEntitlementsFreshInstallIntegrationTest.kt:27]: 👍🏻

**Review by NachoSoto — APPROVED** [review]
